Fix ES module import issues

Muhammad-Altabba opened the pull request:
Fix ES module import issues
+ Add pre-build script to automate the step of adding .js and /index.js to local imports. (once this PR approved the updated files will be pushed).
+ Add script to assign `type: "module"` to package.json at the ES folder.
+ Add and organize the scripts to run before and after build
+ Update some imports of ethers objects to be es compatible

To not make noise with the edited files, once approved or requested will run `pnpm build` which should then:
+ Modify source files to have imports compatible with ES modules.

**Helpful reading:** https://www.embedthis.com/blog/sensedeep/how-to-create-single-source-npm-module.html

**Type of PR:**

- [x] Bugfix
- [ ] Feature
- [ ] Documentation
- [ ] Other

**Required reviews:**

- [ ] 1
- [ ] 2
- [x] 3

**What this does:**

> High-level idea of the changes introduced in this PR. List relevant API
> changes (if any), as well as related PRs and issues.

**Issues fixed/closed:**

Fixes https://github.com/nucypher/sprints/issues/229

**Why it's needed:**

> Explain how this PR fits in the greater context of the NuCypher Network. E.g.,
> if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

> What should reviewers focus on? Is there a particular commit/function/section
> of your PR that requires more attention from reviewers?

manumonti commented on full-check-like-in-ci.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Can we move this script into the `scripts` folder?

manumonti commented on scripts/fix-source-imports.ts:
```diff
@@ -0,0 +1,97 @@
+
+      // Fix all relative import/export statements to include .js extension
+      const importExportRegex =
+        /((?:import|export)(?:\s+type)?\s*(?:{[^}]*}\s*)?(?:(?:\*\s*as\s+\w+|\w+)\s*)?from\s+['"])([^'"]*?)(['"])/g;
```
...but the language is that of Mordor, which I will not utter here. 🤣

manumonti reviewed (changes requested):
Looks good! But something that is not clear to me: what was the problem before this PR? We could use ES modules before, right? What was broken? Why is it needed to add the `.js` extension to the imports?

Muhammad-Altabba commented:
> Looks good! But something that is not clear to me: what was the problem before this PR? We could use ES modules before, right? What was broken? Why is it needed to add the `.js` extension to the imports?

Thanks for your question. Actually the issue was when an ECMA Script project is importing taco-web packages. Sorry I missed initially pointing to the original issue in the PR description. More details on the issue is at: 
 https://github.com/nucypher/sprints/issues/229

manumonti reviewed (approved):
LGTM! 🙌

Muhammad-Altabba commented:
To not cause noise by the edited files (adding .js), once this PR is approved or if requested, I would run `pnpm build` which should then push a commit that:
- Modify source files to have imports compatible with ES modules.

Also by checking out this branch, when executing `pnpm build` the files will be edited locally. Let me know if I should push this commit or not yet.

derekpierre commented:
Seems like this PR needs a rebase...

derekpierre commented:
I'm less familiar with the nuances of ES, CommonJS, Typescript.

Some questions:
* Seems like you are trying to split this PR into two parts:
  1. package.json/script changes
  2. updated imports to use .js extensions etc.

    I think you can employ either of the following strategies:
       1. Do the two changes in this PR as distinct commits, and then we can review commit-by-commit
       OR
       2. this functionality could be implemented in it's own branch (based over `epic-viem`), using separate PRs for each stage, and that branch then gets merged into `epic-viem` once fully completed. That would prevent a half-finished feature from merging into the `epic-viem` branch. 
       
   wdyt?

* I'm interested in how this affects the existing `taco` demos we have in `examples/taco`. The ones that use nodejs, nextjs, react, webpack-5? Do those continue to work as they did before without any changes? If not, does that imply we have backwards compatibility issues with anyone who used `nucypher/taco` in the same way as the demos? Trying to gauge if this is a breaking change, and if so, what exactly it breaks i.e. what does this change mean for already existing usage of `taco-web`?

* Are there any npm publishing/packaging changes we need to make due to this change?

Muhammad-Altabba committed:
chore: rename build:module script to build:es and add "types" field to all package.json files to expose TypeScript declarations

Muhammad-Altabba committed:
feat: add prebuild script that ensure .js extensions in imports

Muhammad-Altabba committed:
feat: add postbuild step to ensure ES module type in dist/es/package.json

Muhammad-Altabba committed:
refactor: update imports to use ethers.utils instead of direct utils import
This is because import from 'ethers/lib/utils' causes compatibility issues with projects that uses ES Modules.

Muhammad-Altabba commented:
Thanks @derekpierre ,
- I just split the PR into 4 small commits for easier review. 
- There must be no breaking changes. 
- And there must be no changes needed for publishing process.

However, in the meantime, I will iterate once more on the last 2 points.

Muhammad-Altabba commented:
Hi @derekpierre,
I just double checked. 

All examples runs with no issues. 
And there must be no breaking changes to any project. Except in a case that should not happen. That is if the user at a CommonJS was wrongly referencing (with `import` or `require` a file from the ESM folder with `@nucypher/taco/dist/es/...` but in a CommonJS project. And because the users are not supposed to import or require except in the format `@nucypher/taco`. And because if the users were ever referencing the `.../dist/...`, in a CommonJS module project, they would reference `.../dist/cjs/...` and not `.../dist/es/...`, if ever. I think no need to raise the library major version for this. Or what do you think?

And there should be no need to modify the publishing process.

theref commented:
ok, in theory i approve this, but I'd like to see the changes that it will make :) @Muhammad-Altabba could you open a PR into this branch that contains: 1 commit with the result of the `pnpm build`

Muhammad-Altabba committed:
add js extension to files import

Muhammad-Altabba commented:
> ok, in theory i approve this, but I'd like to see the changes that it will make :) @Muhammad-Altabba could you open a PR into this branch that contains: 1 commit with the result of the `pnpm build`

https://github.com/nucypher/taco-web/pull/724
Thanks,

theref reviewed (approved):
I'm happy with this once https://github.com/nucypher/taco-web/pull/724 is also included :)
